src/librustc/middle/free_region.rs

```diff
@@ -99,7 +90,26 @@ impl FreeRegionMap {
     /// it is possible that `sub != sup` and `sub <= sup` and `sup <= sub`
     /// (that is, the user can give two different names to the same lifetime).
     pub fn sub_free_region(&self, sub: FreeRegion, sup: FreeRegion) -> bool {
-        can_reach(&self.map, sub, sup) || self.is_static(&sup)
+        let result = sub == sup || {
+            let sub = ty::ReFree(sub);
+            let sup = ty::ReFree(sup);
+            self.relation.contains(&sub, &sup) || self.relation.contains(&sup, &ty::ReStatic)
```

Comment:
@pnkfelix 

> The right-hand disjunct, self.relation.contains(&sup, &ty::ReStatic), does not seem right to me.

Indeed, you'd think I should be able to make some sort of test that will fail here... or at least fail to compile when expected. I think it doesn't make everything go completely nuts simply because ty::ReStatic is usually not in the relation (and when it is, it never has incoming edges).